### DISCUSSION TITLE
Weird DKI output

### DISCUSSION THREAD

**Discussion by araikes** [dipy/dipy#2164] [Reconstruction]
## Description
I'm hoping to get some help here. I have diffusion data that has been pre-processed using QSIPrep (includes bias correction, denoising, Gibbs removal, and eddy). The data were acquired as two separate acquisitions ( _b_ 1000 with 25 directions; _b_ 2000 with 20 directions; two _b_ 0s) that were concatenated as part of the QSIPrep pre-processing. I was extracting tensors from the data. The DTI tensors fit using the DKI model look ok but the DKI maps look strange to me (attached below). 

Do I just have too few directions/shells to actually fit the DKI model?

![dti_tensors](https://user-images.githubusercontent.com/18663624/81208468-ef824080-8f83-11ea-838c-4fdd7dc6ccb2.png)
![dki_tensors](https://user-images.githubusercontent.com/18663624/81208471-f01ad700-8f83-11ea-911d-750ff66bdef7.png)

Code: 
```python
# Import necessary packages
import numpy as np
import glob, os

# BIDS imports
import bids

#DIPY imports
from dipy.io.image import load_nifti, save_nifti
from dipy.io.gradients import read_bvals_bvecs
from dipy.core.gradients import gradient_table

from dipy.segment.mask import median_otsu

import dipy.reconst.dti as dti
import dipy.reconst.dki as dki


# Set base path
data_dir='D:/Documents/GitHub/allo_git/data/neuroimaging'

# Get BIDS layout
layout = bids.layout.BIDSLayout(root = os.path.join(data_dir, 'nifti'),
                                derivatives = os.path.join(data_dir, 
                                                           'derivatives/qsiprep'))

subjects = layout.get_subjects()

subject="allo102"
sesh="bline"

fdwi = layout.get(subject=subject, session = sesh,
               desc="preproc", suffix = 'dwi',
               return_type = 'filename', extension = "nii.gz")

fbvec = layout.get(subject=subject, session = sesh,
               desc="preproc", suffix = 'dwi',
               return_type = 'filename', extension = "bvec")

fbval = layout.get(subject=subject, session = sesh,
               desc="preproc", suffix = 'dwi',
               return_type = 'filename', extension = "bval")
        
# Load data, bvec, and bval files. Create a gradient table object.
data, affine = load_nifti(fdwi[0])
bvals, bvecs = read_bvals_bvecs(fbval[0], fbvec[0])
gtab = gradient_table(bvals, bvecs)

# Create a mask from the data.
masked_data, mask = median_otsu(data, vol_idx = [0,1], median_radius = 4, 
                                numpass = 2, autocrop = True, dilate = 1)

# Fit the DKI model
dki_model = dki.DiffusionKurtosisModel(gtab)
dki_tenfit = dki_model.fit(masked_data)

# Get scalar maps from the DKI model
dki_fa = dki_tenfit.fa
dki_md = dki_tenfit.md
dki_rd = dki_tenfit.rd
dki_ad = dki_tenfit.ad

dki_mk = dki_tenfit.mk(0, 3)
dki_ak = dki_tenfit.ak(0, 3)
dki_rk = dki_tenfit.rk(0, 3)
dki_mkt = dki_tenfit.mkt(0, 3)
dki_kfa = dki_tenfit.kfa
``` 

## Way to reproduce
[If reporting a bug, please include the following important information:]
- [ ] Operating system and version: Windows-10-10.0.18362-SP0
- [ ] Python version 3.7.6
- [ ] dipy version 1.1.0
- [ ] dependency version (numpy, scipy, nibabel, h5py, cvxpy, fury)
    * NumPy: 1.18.1
    * SciPy 1.4.1
    * Nibabel 3.0.2
    * H5py 2.10.0
    * Cvxpy not installed
    * fury 0.5.1

**Comment by skoudoro** [comment]
Hi @araikes,

Thank you for your question/feedback, I will just tag @RafaelNH @arokem @Garyfallidis @ShreyasFadnavis @mdesco who can address this question because of their experience.

**Comment by araikes** [comment]
Thanks @skoudoro

**Comment by Garyfallidis** [comment]
@RafaelNH can you please get back to @araikes ?

**Comment by RafaelNH** [comment]
Hi @araikes! I think the issue is coming from the independent acquisitions. When concatonating the data you have to insure that the intensity scales of two diffusion-weighted dataset have the same scale (intensity scales can be different if you have for example different receiver gains of the different acquisitions which is likey to be automatically adjusted by the scanner). To check that issue, see if the b0s intensities vary a lot from acquisition to acquisition.
A post-processing trick to virtual produce same intensity scales is to compute the mean S0 intensity value for each dataset and then devide the intere datated by their mean S0 intensity values. This is not an optimal procedure but may overcome the issue if you don't have the information of the receiver gain.
Another solution is to compute the signal decays (S/mean(S0)) voxel by voxel before concatonating and before inputing the data to the dki fitting procedures. For that to work, you also have to give the b0 information - i.e. S0/mean(S0) which is expected to be near 1 (deviations around 1 are due to thermal noise).

**Comment by araikes** [comment]
Hi @RafaelNH ,
Thanks for the input. I'm curious about one thing (among many, I suppose). When you say that the intensity scales have to have the same scale, does this suggest that the b0s need to "match" or simply that they need to be reasonably close? For example, for the individual in the original picture, the b0s from each acquisition are pasted below. Further, the QSIPrep processing steps run denoising before concatenating and is supposed to b0 harmonization. I'll check on that though.

Regarding your post-processing trick, do you have any ready-to-edit code that would let me try that?

Thanks

![sub-allo102_ses-bline_acq-b1000_dir-PA_run-001_dwi](https://user-images.githubusercontent.com/18663624/81846731-692ba880-9507-11ea-817a-be14660f6768.png)
b1000 acquisition

![sub-allo102_ses-bline_acq-b2000_dir-PA_run-001_dwi](https://user-images.githubusercontent.com/18663624/81846743-6df05c80-9507-11ea-9e78-56686ac6debd.png)
b2000 acquisition

**Comment by araikes** [comment]
Potentially related question @RafaelNH: In the FA map from the DKI reconstruction, There are areas with missing voxels where there is more than likely  brain tissue. The DTI reconstructions and Free Water Elimination models all of have voxel values for these but the DKI gives a 0. 

Any thoughts? Same issue as above?

![dki_fa](https://user-images.githubusercontent.com/18663624/81865161-6a1e0380-9522-11ea-85bb-86823a901127.png)
FA map from the DKI reconstruction. There is an obvious patch of voxels with 0 intensity in the middle of an area that appears to be white matter.

![full_fa](https://user-images.githubusercontent.com/18663624/81865200-72763e80-9522-11ea-891c-855d4afcf012.png)
FA map from the DTI reconstruction using all shells (b0, b1000, b2000). That hole is filled in.

![threshed_fa](https://user-images.githubusercontent.com/18663624/81865261-80c45a80-9522-11ea-990a-ef8bdf0a71dc.png)
FA map from Free Water Elimination reconstruction (with free water volume > 0.7 = 0 as in the tutorial). That whole from the DKI fit is filled here as well. Note that this is a cohort with Alzheimer's so some of other zeroed out values may reflect cortical atrophy

**Comment by marcobattiston1988** [comment]
Hi all,

Sorry for jumping in this conversation, but I am experiencing similar issues to those reported here.

In our lab, we are using dipy in most of our analysis, and it is a great tool indeed (thanks a lot for developing this and making it available to anyone!!!), however we are struggling a bit to get satisfactory DKI maps: that known "black voxels" issue mention above here.

My little knowledge tells me this is not an issue of dipy itself, but it is a more general problem in DKI model fitting (number of parameters, low SNR, sensitivity to physiological noise, etc.), but I believe that with a little bit of tweaking the fitting can be improved.

I attach here an example of MK from spinal cord data (9 b0, 64 dw volumes as b=1000s/mm2 (22), b=2000s/mm2 (21), b=3000s/mm2 (21)). On the left, the output from a standard configuration of dipy (i.e. basically a copy/past from your tutorial), on the right the tool we were previously using (developed in Matlab, available here [https://cai2r.net/resources/software/diffusion-kurtosis-imaging-matlab-toolbox](url))
![crop1-d](https://user-images.githubusercontent.com/29372400/95568999-9e441480-0a2d-11eb-9af3-00e9e2bfcb40.png)![crop1-m](https://user-images.githubusercontent.com/29372400/95569033-abf99a00-0a2d-11eb-9321-0c01b2bfd38a.png)
![crop2-d](https://user-images.githubusercontent.com/29372400/95569121-cf244980-0a2d-11eb-83d7-015a330c6543.png)![crop2-m](https://user-images.githubusercontent.com/29372400/95569127-d21f3a00-0a2d-11eb-82c2-3a756286d37c.png)

I did a couple of quick tests, by changing the minimizer function (from wls to ols), tried also the restore method (but I couldn't get it to work properly). I have seen that there are quite a few stuff implemented there in the code, maybe you guys could point me toward some alternative ways to fit these data? 

If there is an interest in investigating this further, we have plenty of data that (I believe ?!) I could share.
